### PR TITLE
Optimize `max_boolean` by operating on u64 chunks

### PR DESCRIPTION
**Pull request by simonvandel**
# Which issue does this PR close?

Closes #.

# Rationale for this change

Vectorizing boolean operations using u64 chunks is generally more efficient.

# What changes are included in this PR?

- Adds benchmarks for boolean min/max
- Operate on bit chunks instead of individual booleans, which can lead to massive speedups while not regressing the short-circuiting behavior of the existing implementation.

`cargo bench --bench aggregate_kernels -- "bool/max"` shows throughput improvements between 50% to 23390% on my machine.

<details><summary>Full benchmark results:</summary>

```
cargo bench --bench aggregate_kernels -- "bool/max"
bool/max nonnull mixed  time:   [4.9617 ns 4.9934 ns 5.0338 ns]
                        thrpt:  [ 13019 Gelem/s  13125 Gelem/s  13208 Gelem/s]
                 change:
                        time:   [-36.288% -34.043% -31.678%] (p = 0.00 < 0.05)
                        thrpt:  [+46.365% +51.614% +56.957%]
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) high mild
  8 (8.00%) high severe
bool/max nonnull false  time:   [565.07 ns 588.68 ns 611.85 ns]
                        thrpt:  [107.11 Gelem/s 111.33 Gelem/s 115.98 Gelem/s]
                 change:
                        time:   [-98.893% -98.849% -98.802%] (p = 0.00 < 0.05)
                        thrpt:  [+8243.8% +8589.5% +8936.8%]
                        Performance has improved.
bool/max nonnull true   time:   [5.0621 ns 5.0978 ns 5.1435 ns]
                        thrpt:  [ 12741 Gelem/s  12856 Gelem/s  12946 Gelem/s]
                 change:
                        time:   [-4.7786% -1.8889% +1.1654%] (p = 0.23 > 0.05)
                        thrpt:  [-1.1520% +1.9253% +5.0184%]
                        No change in performance detected.
Found 10 outliers among 100 measurements (10.00%)
  3 (3.00%) high mild
  7 (7.00%) high severe
bool/max nullable mixed time:   [9.4018 ns 9.5263 ns 9.6842 ns]
                        thrpt:  [6767.3 Gelem/s 6879.5 Gelem/s 6970.6 Gelem/s]
                 change:
                        time:   [-41.198% -38.405% -35.720%] (p = 0.00 < 0.05)
                        thrpt:  [+55.568% +62.350% +70.063%]
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  3 (3.00%) high mild
  10 (10.00%) high severe
bool/max nullable false time:   [1.1109 µs 1.1591 µs 1.2086 µs]
                        thrpt:  [54.223 Gelem/s 56.539 Gelem/s 58.994 Gelem/s]
                 change:
                        time:   [-99.592% -99.574% -99.556%] (p = 0.00 < 0.05)
                        thrpt:  [+22422% +23392% +24407%]
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
bool/max nullable true  time:   [9.0433 ns 9.0830 ns 9.1284 ns]
                        thrpt:  [7179.4 Gelem/s 7215.3 Gelem/s 7246.9 Gelem/s]
                 change:
                        time:   [-39.120% -37.104% -34.921%] (p = 0.00 < 0.05)
                        thrpt:  [+53.659% +58.992% +64.257%]
                        Performance has improved.
Found 11 outliers among 100 measurements (11.00%)
  1 (1.00%) high mild
  10 (10.00%) high severe
```

</details>

# Are there any user-facing changes?

Faster `max_boolean`.

### CODE REVIEW

File: arrow-arith/src/aggregate.rs [hunk 1]
```diff
@@ -371,11 +371,26 @@ pub fn max_boolean(array: &BooleanArray) -> Option<bool> {
...
-        .flatten()
-        .or(Some(false))
+    match array.nulls() {
+        None => array
+            .values()
+            .bit_chunks()
+            .iter_padded()
+            // We found a true if any bit is set
+            .map(|x| x != 0)
+            .find(|b| *b)
```
Comment (Xuanwo) [inline arrow-arith/src/aggregate.rs:381]: Is it a good idea to use `find_map` instead?

File: arrow-arith/src/aggregate.rs [hunk 2]
```diff
@@ -371,11 +371,26 @@ pub fn max_boolean(array: &BooleanArray) -> Option<bool> {
...
+            .or(Some(false)),
+        Some(nulls) => {
+            let validity_chunks = nulls.inner().bit_chunks().iter_padded();
+            let value_chunks = array.values().bit_chunks().iter_padded();
+            value_chunks
+                .zip(validity_chunks)
+                // We found a true if the value bit is 1, AND the validity bit is 1 for any bits in the chunk
+                .map(|(value_bits, validity_bits)| (value_bits & validity_bits) != 0)
```
Comment (Xuanwo) [inline arrow-arith/src/aggregate.rs:389]: The same question.

**Review by Xuanwo — COMMENTED** [review]
Perfect. This PR looks good to me overall, and the newly added bench is really great. I just have some little questions.

File: arrow-arith/src/aggregate.rs (hunk 1 above)
Comment (simonvandel) [reply arrow-arith/src/aggregate.rs:381]: Do you mean like this? `.find_map(|x| if x != 0 { Some(true) } else { None })`

It's a personal preference, but I prefer the existing map + find

File: arrow-arith/src/aggregate.rs [hunk 3]
```diff
@@ -371,11 +371,26 @@ pub fn max_boolean(array: &BooleanArray) -> Option<bool> {
...
-        .find(|&b| b == Some(true))
-        .flatten()
-        .or(Some(false))
+    match array.nulls() {
+        None => array
+            .values()
+            .bit_chunks()
+            .iter_padded()
+            // We found a true if any bit is set
+            .map(|x| x != 0)
```
Comment (XiangpengHao) [inline arrow-arith/src/aggregate.rs:380]: Very cool! I wonder if we can do similar to `min_boolean` where the condition is `.map(|x| x != u64::MAX)`?

Comment (simonvandel) [reply arrow-arith/src/aggregate.rs:380]: Absolutely! I think I'll do it in a follow-up PR

**Comment by alamb** [comment]
>  23390%

That is quite an improvement

File: arrow-arith/src/aggregate.rs [hunk 4]
```diff
@@ -1318,6 +1333,14 @@ mod tests {
         let a = BooleanArray::from(vec![Some(false), Some(true), None, Some(false), None]);
         assert_eq!(Some(false), min_boolean(&a));
         assert_eq!(Some(true), max_boolean(&a));
+
```
Comment (alamb) [inline arrow-arith/src/aggregate.rs:1336]: Can we please add some tests that have more than 64 values to exercise this new code?

I am thinking:
1. all true in the first 64 bits, false in the second
2.  all false in the first 64 bits, true in the second
3. all true (more than 64 bits, less than 128)
4. all false (more than 64 its, less than 128)

Both with/without nulls

**Review by alamb — COMMENTED** [review]
Thanks @simonvandel  and @Xuanwo -- this looks pretty neat to me

I think it just needs a few more tests and it will be good to go

File: arrow-arith/src/aggregate.rs (hunk 4 above)
Comment (simonvandel) [reply arrow-arith/src/aggregate.rs:1337]: Thanks for the feedback.
I added some tests in 5c40f28ec4716e65b9802e8d06941d2f093e3b8c.
Is this what you were thinking of?

**Review by alamb — APPROVED** [review]
Looks good to me -- thank you @simonvandel

File: arrow-arith/src/aggregate.rs (hunk 1 above)
Comment (Xuanwo) [reply arrow-arith/src/aggregate.rs:381]: I'm thinking about `(x!=0).then_some(true)`. But I do agree that it's just personal preference.

**Review by Dandandan — APPROVED** [review]
Nice!

**Pull request merged**